src/accumulator/proof.rs

```diff
@@ -241,6 +243,291 @@ impl Proof {
+            add_hashes,
+            cached_hashes.clone(),
+            remembers,
+            update_data.new_add,
+            update_data.prev_num_leaves,
+            update_data.to_destroy,
+        )?;
+
+        Ok(data_after_addition)
+    }
+    fn update_proof_add(
+        self,
+        adds: Vec<sha256::Hash>,
+        cached_del_hashes: Vec<sha256::Hash>,
+        remembers: Vec<u64>,
+        new_nodes: Vec<(u64, sha256::Hash)>,
+        before_num_leaves: u64,
+        to_destroy: Vec<u64>,
+    ) -> Result<(Proof, Vec<sha256::Hash>), String> {
+        // Combine the hashes with the targets.
+        let orig_targets_with_hash: Vec<(u64, sha256::Hash)> = self
+            .targets
+            .to_owned()
+            .into_iter()
+            .zip(cached_del_hashes.into_iter())
+            .collect();
+
+        // Attach positions to the proof.
+        let proof_pos = get_proof_positions(
+            &self.targets,
+            before_num_leaves,
+            util::tree_rows(before_num_leaves),
+        );
+        let proof_with_pos = proof_pos
+            .clone()
+            .into_iter()
+            .zip(self.hashes.clone())
+            .collect();
+
+        // Remap the positions if we moved up a after the addition row.
+        let targets_after_remap = Proof::maybe_remap(
+            before_num_leaves,
+            adds.len() as u64,
+            orig_targets_with_hash.clone(),
+        );
+        let mut final_targets = targets_after_remap.clone();
+        let mut new_nodes_iter = new_nodes.iter();
+        let proof_with_pos =
+            Proof::maybe_remap(before_num_leaves, adds.len() as u64, proof_with_pos);
+        // remembers is an index telling what newly created UTXO should be cached
+        for remember in remembers {
+            let remember_target: Option<&sha256::Hash> = adds.get(remember as usize);
+            if let Some(remember_target) = remember_target {
+                let node = new_nodes_iter.find(|(_, hash)| *hash == *remember_target);
+                if let Some((pos, hash)) = node {
+                    final_targets.push((*pos, *hash));
+                }
+            }
+        }
+
+        final_targets.sort();
+        // Move up positions that need to be moved up due to the empty roots
+        // being written over.
+        let num_leaves = before_num_leaves + (adds.len() as u64);
+        let (mut new_target_pos, target_hashes): (Vec<_>, Vec<_>) =
+            final_targets.clone().into_iter().unzip();
+        let mut new_proof_pos = proof_pos;
+
+        for node in to_destroy {
+            (new_target_pos, _) =
+                Proof::calc_next_positions(&vec![node], &final_targets, num_leaves, true)?;
+            (new_proof_pos, _) =
+                Proof::calc_next_positions(&vec![node], &proof_with_pos, num_leaves, true)?;
+        }
+        // Add in proof_hashes to the new_nodes as some needed hashes
```

Comment:
Where are the proof_hashes being added in? I don't see it in the code

Reply:
Do you mean in the new proof?

Reply:
I mean the old proof hashes (that we've updated with new positions).

In the go impl I create a completely new proof object so you need to also grab the old proof hashes as well. Looks like you're doing that part in a different way.

The comment isn't really describing what the code is doing so it should be removed.